### PR TITLE
add Mfp3Encoder

### PR DESCRIPTION
**Pull request by thiloSchlemmer**

### CODE REVIEW

File: src/modules/olMap/formats/Mfp3Encoder.js [hunk 1]
```diff
@@ -0,0 +1,599 @@
...
+				thirdPartyDataOwner: encodedLayers.thirdPartyDataOwners.length !== 0 ? encodedLayers.thirdPartyDataOwners.join(',') : '',
+				shortLink: shortLinkUrl,
+				qrcodeurl: qrCodeUrl
+			}
+		};
+	}
+
+	_geoResourceFrom(layer) {
+		const layerId = layer.get('id');
+		const geoResourceId = layer.get('geoResourceId');
```
Comment (taulinger) [inline src/modules/olMap/formats/Mfp3Encoder.js:139]: presence of `geoResourceId` should be guaranteed

Comment (thiloSchlemmer) [reply src/modules/olMap/formats/Mfp3Encoder.js:139]: Currently only VectorLayers have a `geoResourceId`. For WMTS- and WMS-Georesources the binding to a Layer is not done yet ([see LayerService.js  Line 83-97](https://github.com/ldbv-by/bav4-nomigration/blob/0e6e32ce157c1411c28d73a704324e8696d3c766/src/modules/olMap/services/LayerService.js#L89))

Comment (taulinger) [reply src/modules/olMap/formats/Mfp3Encoder.js]: presence of `geoResourceId` should be guaranteed

Comment (taulinger) [reply src/modules/olMap/formats/Mfp3Encoder.js]: Thx, your're right

**Review by taulinger — APPROVED** [review]
Let's go 🥳

File: src/modules/olMap/formats/Mfp3Encoder.js [hunk 2]
```diff
@@ -0,0 +1,684 @@
...
+const PixelSizeInMeter = 0.00028; // based on https://www.adv-online.de/AdV-Produkte/Standards-und-Produktblaetter/AdV-Profile/binarywriterservlet?imgUid=36060b99-b8c4-0a41-ba3c-cdd1072e13d6&uBasVariant=11111111-1111-1111-1111-111111111111  and the calculations of a specific scaleDenominator (p.22)
+
+/**
+ * Encoder to create a MapFishPrint request-object from a OpenLayers map-instance
+ * @author thiloSchlemmer
+ * @interface Mfp3Encoder
+ */
+
+/**
+ * Encodes the content of a OpenLayers {@see Map} to a MapFishPrint3 job request.
```
Comment (taulinger) [inline src/modules/olMap/formats/Mfp3Encoder.js:29]: `Encodes the content of a ol {@see Map} to a MapFishPrint3 job request.`

File: src/modules/olMap/formats/Mfp3Encoder.js [hunk 3]
```diff
@@ -0,0 +1,684 @@
...
+ * @function
+ * @async
+ * @name Mfp3Encoder#encode
+ * @param {ol.Map} olMap the map with the content to encode for MapFishPrint3
```
Comment (taulinger) [inline src/modules/olMap/formats/Mfp3Encoder.js:33]: just `ol map` or `ol {@see Map}` ?

File: src/modules/olMap/formats/Mfp3Encoder.js [hunk 4]
```diff
@@ -0,0 +1,684 @@
...
+ * @param {EncodingProperties} encodingProperties the map with the content to encode for MapFishPrint3
```
Comment (taulinger) [inline src/modules/olMap/formats/Mfp3Encoder.js:34]: just `{EncodingProperties} encodingProperties` ?

File: src/modules/olMap/handler/mfp/OlMfpHandler.js [hunk 5]
```diff
@@ -222,4 +228,14 @@ export class OlMfpHandler extends OlLayerHandler {
...
 		return angle;
 	}
+
+	async _encodeMap() {
+		const { id, scale, dpi } = this._storeService.getStore().getState().mfp.current;
+		const pageCenter = this._getVisibleCenterPoint();
+		const encodingProperties = { layoutId: id, scale: scale, rotation: 0, dpi: dpi, pageCenter: pageCenter };
+		const specs = await this._encoder.encode(this._map, encodingProperties);
+
+		// console.log(specs);
```
Comment (taulinger) [inline src/modules/olMap/handler/mfp/OlMfpHandler.js:238]: let's remove this line 😊

File: test/modules/olMap/formats/Mfp3Encoder.test.js [hunk 6]
```diff
@@ -0,0 +1,1332 @@
...
+import { fromLonLat } from 'ol/proj';
+import { WMTS, XYZ } from 'ol/source';
+import TileLayer from 'ol/layer/Tile';
+import WMTSTileGrid from 'ol/tilegrid/WMTS';
+import LayerGroup from 'ol/layer/Group';
+import TileGrid from 'ol/tilegrid/TileGrid';
+import { AdvWmtsTileGrid } from '../../../../src/modules/olMap/ol/tileGrid/AdvWmtsTileGrid';
+import { MeasurementOverlayTypes } from '../../../../src/modules/olMap/components/MeasurementOverlay';
+
+describe('Mfp3Encoder', () => {
```
Comment (taulinger) [inline test/modules/olMap/formats/Mfp3Encoder.test.js:28]: `'BvvMfp3Encoder'`

**Review by taulinger — COMMENTED** [review]
Nice!
Would suggest moving `Mfp3Encoder.js` from /formats to /services.

**Review by taulinger — APPROVED** [review]
Yippie!

**Pull request merged**